### PR TITLE
Parrying Refactor, Adding Parrying to Select Two-handed Weapons: Parry This You Filthy Casual!

### PR DESCRIPTION
**Pull request by Spaghetti-bit**
## What Does This PR Do

Refactors parrying and adds conditions for parrying, this affects two-handed weapons mostly. All two handed weapons (if they can parry) will be required to be wielded in both hands before they can parry.

Items like the energy sword and riot shield make be activated before they can parry.

The spear, fireaxe, wizard melee weapons, and most two-handed items have been given some sort of parry (albeit some of these parries are not the best when it comes to uptime, this is intentional). 

Higher power items like wizard items can parry projectiles while stuff like the fireaxe and only parry melee attacks.

Yes, you can parry with push broom. AND the titanium push broom.

## Why It's Good For The Game

Being able to parry with an energy sword that is off is not intentional. 
Being able to parry with telescopic riot shield while it is not extended is not intended.

A refactor of this mechanic allows for further implementation down the line and adding new items that can parry. Parrying is a sadly underused mechanic and adds a layer of skill to combat while also not breaking the balance of the game.

## Images of changes

N/A

## Testing

Bullied blooddrunk miner with a push broom for about 10 seconds and then got folded like laundry. (It could parry)
Tried to parry with a tele shield without extending it, cried! Extended it and giggled! (It could parry) 
Tried to parry with an e-sword with it off! Wow! Cried! Turned it on and smiled! (It could parry)
Tested all two handed weapons that can parry while holding it with only one hand and felt weak! Held it in both hands and felt powerful! (It could parry)

## Changelog
:cl:
tweak: Parrying with two handed weapons now requires you to wield them.
tweak: Parrying with the energy sword and telescopic riot shield now require them to be active.
experimental: Added parrying to select two-handed weapons.
/:cl:

### CODE REVIEW

File: code/game/objects/items/weapons/melee/energy_melee_weapons.dm
Comment (DGamerL) [inline code/game/objects/items/weapons/melee/energy_melee_weapons.dm:52]:
```suggestion
		REMOVE_TRAIT(src, TRAIT_ITEM_ACTIVE, TRAIT_GENERIC)
```

File: code/game/objects/items/weapons/melee/energy_melee_weapons.dm
Comment (DGamerL) [inline code/game/objects/items/weapons/melee/energy_melee_weapons.dm:346]:
```suggestion
		REMOVE_TRAIT(src, TRAIT_ITEM_ACTIVE, TRAIT_GENERIC)
```

**Comment by 1080pCat** [comment]
> Yes, you can parry with push broom.

hell yea

File: code/datums/components/parry.dm
Comment (lewcc) [inline code/datums/components/parry.dm:84]:
```suggestion
	if(requires_two_hands && !HAS_TRAIT(parent, TRAIT_WIELDED)) // If our item has special conditions before being able to parry.
		return
```

File: code/game/objects/items/weapons/melee/energy_melee_weapons.dm
Comment (lewcc) [inline code/game/objects/items/weapons/melee/energy_melee_weapons.dm:281]: this one is always active, yeah?
```suggestion
	ADD_TRAIT(src, TRAIT_ITEM_ACTIVE, ROUNDSTART_TRAIT)
```

File: code/game/objects/items/weapons/melee/energy_melee_weapons.dm
Comment (lewcc) [inline code/game/objects/items/weapons/melee/energy_melee_weapons.dm:363]: can you restore this? the different clanging sounds are pretty iconic for this item.

**Review by lewcc — APPROVED** [review]
Code LGTM, pending approval

**Comment by Coolrune206** [comment]
Push broom deflects ranged attacks already and should not also get a 75% uptime melee parry.

**Comment by Spaghetti-bit** [comment]
> Push broom deflects ranged attacks already and should not also get a 75% uptime melee parry.

It costs a large amount of TC, if you don't have anti-drop the item will 90% just be lost from a wall splat. Giving it a high-uptime of parrying puts it on par with carp.

**Comment by Fordoxia** [comment]
More parry is more good.

File: code/game/objects/items/weapons/melee/energy_melee_weapons.dm
Comment (hal9000PR) [inline code/game/objects/items/weapons/melee/energy_melee_weapons.dm:325]:
```suggestion
		if(!HAS_TRAIT(src, TRAIT_ITEM_ACTIVE)) //This isn't in the above if-check so that the else doesn't care about active
```
surely we should be checking if it is inactive? like the old code used to?

**Review by hal9000PR — CHANGES_REQUESTED** [review]
some comments

File: code/game/objects/items/weapons/shields.dm
Comment (hal9000PR) [inline code/game/objects/items/weapons/shields.dm:169]: src is implied.

**Review by hal9000PR — APPROVED** [review]
remove that one `src.` and this should be ready to merge

File: code/game/objects/items/weapons/shields.dm
Comment (DGamerL) [reply code/game/objects/items/weapons/shields.dm:169]:
```suggestion
	playsound(loc, 'sound/weapons/batonextend.ogg', 50, TRUE)
```

**Comment by Spaghetti-bit** [comment]
Use to C sharp and C++, apologies... Haha! `This.` and whatnot <3

**Review by hal9000PR — APPROVED** [review]

**Review by DGamerL — APPROVED** [review]

**Pull request merged**
